Sources/MongoSwift/MongoCollection+FindAndModify.swift

```diff
@@ -50,32 +56,61 @@ extension MongoCollection {
      *
      * - Returns: A `CollectionType` representing either the original or updated document,
      *      depending on selected options, or `nil` if there was no match.
-     * - Throws: 
+     *
+     * - Throws:
+     *   - `MongoError.invalidArgument` if any of the provided options are invalid
      *   - `MongoError.commandError` if there are any errors executing the command.
      *   - A `DecodingError` if the updated document cannot be decoded to a `CollectionType` value
      */
     @discardableResult
     public func findOneAndUpdate(filter: Document, update: Document,
                                  options: FindOneAndUpdateOptions? = nil) throws -> CollectionType? {
-        throw MongoError.commandError(message: "Unimplemented command")
+        return try self.findAndModify(filter: filter, update: update, options: options)
+    }
+
+    /// A private helper method for findAndModify operations to use
+    private func findAndModify(filter: Document, update: Document? = nil,
+                               options: FindAndModifyOptionsConvertible? = nil) throws -> CollectionType? {
+
+        var opts = try options?.asOpts()
+
+        if let update = update {
+            // if no options were provided but we need to set update, create them
+            opts = opts ?? FindAndModifyOptions()
+            try opts!.setUpdate(update)
```

Comment:
Is `opts!` necessary here? I would expect the null coalescing on the previous line guarantees that it's an object (also since `FindAndModifyOptions.init` doesn't throw).

Reply:
yes, since `opts` is initially an optional type, it remains that way.